### PR TITLE
Add filter methods for data object fields in their corresponding listing classes

### LINKED ISSUE #4351: Filter methods for data object listings
Currently the listing classes only extend \Pimcore\Model\DataObject\Listing\Concrete and set some class properties.
What do you think of adding filter methods for the class fields like 
```php
class Listing extends DataObject\Listing\Concrete {
...
  public function filterByName($name) {
    $this->addConditionParam("`name` = ?", $name);
  }
...
}

$userListing = new Listing();
$userListing->filterByName($name);
// $userListing now contains all User data objects whose field name=$name
```
Of course this could be done via `User::getByName($name);` but these methods only support filtering by one field.

Or even add comparison operators so you could do
```php
$userListing = new Listing();
// get all users who registered since 2019-01-01 00:00:00
$userListing->filterByRegistrationDate(new DateTime('2019-01-01 00:00:00", ">=");
```

Actually I am inly interested if this feature sounds interesting. I could try to implement a PR if you want.

### PR DESCRIPTION
**Pull request by BlackbitNeueMedien**
This PR adds methods `filterByXYZ()` to generated data object listing classes.
Additionally it adds &lt;Data Object Class&gt;::getByXYZ() magic method call definition only for those fields which allow filtering (till now the getByXYZ magic method got added for every field and if it was not supported there was a runtime exception)
And it adds &lt;Data Object Class&gt;::getByXYZ() for localized fields (not only getByLocalizedFields(...) but also if xyz is a localized field you can do &lt;Data Object Class&gt;::getByXYZ($value, $locale))

Resolves #4351

### CODE REVIEW

**Comment by BlackbitNeueMedien** [comment]
I think this is ready to get reviewed.

File: models/DataObject/Concrete.php [hunk 1]
```diff
@@ -624,13 +634,13 @@ public static function __callStatic($method, $arguments)
                     throw new \Exception('Call to undefined static method ' . $method . ' in class DataObject\\Concrete');
                 }
 
-                if (!in_array($localizedField->getFieldType(), $allowedDataTypes)) {
-                    throw new \Exception("Static getter '::getBy".ucfirst($propertyName)."' is not allowed for fieldtype '" . $localizedField->getFieldType() . "', it's only allowed for the following fieldtypes: " . implode(',', $allowedDataTypes));
+                if ($localizedField->isFilterable()) {
```
Comment (blankse) [inline models/DataObject/Concrete.php]: I think, it should be `if (!$localizedField->isFilterable()) {` right?

Comment (BlackbitNeueMedien) [reply models/DataObject/Concrete.php]: You are right, thanks. Fixed this.

**Comment by dvesh3** [comment]
Hey @BlackbitNeueMedien! sorry for delayed review, missed it somehow. Can you please rebase master?

**Comment by dvesh3** [comment]
@BlackbitNeueMedien Thanks for your updates...are you done with your changes?

**Comment by BlackbitNeueMedien** [comment]
@dvesh3 Yes, I am done. Certainly we could add more data types to be filterable but this could also be done in the future.

File: models/DataObject/Concrete.php [hunk 2]
```diff
@@ -647,13 +646,23 @@ public static function __callStatic($method, $arguments)
         }
         $propertyName = implode('', preg_grep('/^' . preg_quote($propertyName, '/') . '$/i', $fieldnames));
```
Comment (dvesh3) [inline models/DataObject/Concrete.php]: **getByFieldName** not working for localized fields. It throws error "Call to undefined static method getBy...."

File: models/DataObject/ClassDefinition/Data.php [hunk 3]
```diff
@@ -1010,6 +1010,27 @@ public function getSetterCodeLocalizedfields($class)
...
+     *
+     * @return string
+     */
+    public function getFilterCode()
+    {
+        $key = $this->getName();
+
+        $code = '/**' . "\n";
+        $code .= '* Filter by ' . str_replace(['/**', '*/', '//'], '', $key) . ' - ' . str_replace(['/**', '*/', '//'], '', $this->getTitle()) . "\n";
+        $code .= '* @return static'."\n";
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data.php]: params missing in generated doc

File: doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md [hunk 4]
```diff
@@ -96,6 +96,12 @@ $entries->setCondition("city IN (:cities)", ["cities" => ["New York", "Chicago"]
 
 //if necessary you can of course custom build your query
 $entries->setCondition("name LIKE " . $entries->quote("%bernie%")); // make sure that you quote variables in conditions!
+
+// some data types support direct filtering
+$entries->filterByName('Jan'); // filters for name='Jan'
+$entries->filterByAge(18, '>='); // filters for age >= 18
+$entries->filterByCity(['New York', 'Chicago'], 'IN (?)'); // filters for city IN ('New York','Chicago')
```
Comment (dvesh3) [inline doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md]: [["New York", "Chicago"]]

Comment (BlackbitNeueMedien) [reply doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md]: Fixed.

File: models/DataObject/ClassDefinition/Data.php (hunk 3 above)
Comment (BlackbitNeueMedien) [reply models/DataObject/ClassDefinition/Data.php]: Fixed.

File: models/DataObject/Concrete.php (hunk 2 above)
Comment (BlackbitNeueMedien) [reply models/DataObject/Concrete.php]: Fixed.

File: models/DataObject/ClassDefinition/Data.php [hunk 5]
```diff
@@ -1020,7 +1020,21 @@ public function getFilterCode()
...
-        $code .= '* Filter by ' . str_replace(['/**', '*/', '//'], '', $key) . ' - ' . str_replace(['/**', '*/', '//'], '', $this->getTitle()) . "\n";
+        $code .= '* Filter by ' . str_replace(['/**', '*/', '//'], '', $key) . ' (' . str_replace(['/**', '*/', '//'], '', $this->getTitle()) . ")\n";
+
+        $dataParamDoc = 'mixed $'.$key;
+        $reflectionMethod = new \ReflectionMethod($this, 'addListingFilter');
+        if(preg_match('/@param\s+([^\s]+)\s+\$data(.*)/', $reflectionMethod->getDocComment(), $dataParam)) {
+            $dataParamDoc = $dataParam[1].' $'.$key.' '.$dataParam[2];
+        }
+
+        $operatorParamDoc = 'string $operator SQL comparison operator, e.g. =, <, >= etc. You can use "?" as placeholder, e.g. "IN (?)"';
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data.php]: incorrect generated phpdoc
image

File: doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md [hunk 6]
```diff
@@ -100,7 +100,7 @@ $entries->setCondition("name LIKE " . $entries->quote("%bernie%")); // make sure
 // some data types support direct filtering
 $entries->filterByName('Jan'); // filters for name='Jan'
 $entries->filterByAge(18, '>='); // filters for age >= 18
-$entries->filterByCity(['New York', 'Chicago'], 'IN (?)'); // filters for city IN ('New York','Chicago')
+$entries->filterByCity([['New York', 'Chicago']], 'IN (?)'); // filters for city IN ('New York','Chicago')
```
Comment (dvesh3) [inline doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md]: No need to update example...if data wrapped in array!

Comment (BlackbitNeueMedien) [reply doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md]: Ah sorry, I did not want to commit the parameter wrapping in an array to allow the comparison operator to have multiple `?` placeholders.
Now the doc matches the code.

File: models/DataObject/ClassDefinition/Data.php (hunk 5 above)
Comment (BlackbitNeueMedien) [reply models/DataObject/ClassDefinition/Data.php]: Fixed.

File: doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md [hunk 7]
```diff
@@ -238,6 +247,9 @@ foreach ($list as $city) {
 
 ```php
 $list = DataObject\News::getByLocalizedfields($fieldName, $value, $locale, $limit | array($limit, $offset, $unpublished));
+
+// or
+$list = DataObject\News::getByFieldName($value, $locale, $limit | array($limit, $offset, $unpublished));
```
Comment (dvesh3) [inline doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md]: @BlackbitNeueMedien could you please fix the wrong example here for last parameter as array? (AFAIK it is not possible) and also for `getByLocalizedfields`.

Or if possible, changes to support it? Thanks!

Comment (BlackbitNeueMedien) [reply doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md]: Fixed.
Only the example in the docs was wrong. `getByLocalizedfields()` also works with list config as 4th parameter - but it has to be an associative array.

Comment (dvesh3) [reply doc/Development_Documentation/05_Objects/03_Working_with_PHP_API.md]: ohh got it,..😄 thanks for your changes!

**Comment by dvesh3** [comment]
@BlackbitNeueMedien Hello, could you please resolve the conflicts? so we that we can merge it 😉

**Comment by BlackbitNeueMedien** [comment]
@dvesh3 Merged master.

**Comment by dvesh3** [comment]
@BlackbitNeueMedien thanks a lot!

**Pull request merged**